### PR TITLE
[NFC] Replace miopen::ProblemDescription with conv::ProblemDescription, part 2

### PR DESCRIPTION
**Pull request by averinevg**
List of changes:

- `ProblemDescription` data fields have been removed
- `mlo_construct_direct2D_fusion` has been replaced with `ProblemDescription` and removed
- `ProblemDescriptionCompatTemporary` has been created for use in legacy code
- Unused code has been removed

Blocker: https://github.com/ROCmSoftwarePlatform/MIFin/pull/83

See https://github.com/ROCmSoftwarePlatform/MIOpen/issues/266

### CODE REVIEW

File: src/conv/problem_description.cpp [hunk 1]
```diff
@@ -188,6 +194,36 @@ bool ProblemDescription::IsLayoutDefault() const
...
+        return (in_layout == "NDHWC") && (out_layout == "NDHWC") && (weights_layout == "NDHWC");
+    }
+}
+
+bool ProblemDescription::IsLayoutNCHWc() const
+{
+    return ((GetSpatialDims() == 2) && (in_layout == "NCHWc") && (out_layout == "NCHWc") &&
+            (weights_layout == "NCHWc")) ||
+           ((GetSpatialDims() == 2) && (in_layout == "NCHWc") && (out_layout == "NCHWc") &&
+            (weights_layout == "CHWNc"));
```
Comment (atamazov) [inline src/conv/problem_description.cpp]: [R][Quality]
```suggestion
    return IsNCHWc_NCHWc() || IsNCHWc_CHWNc;
```

File: src/include/miopen/conv/problem_description.hpp [hunk 2]
```diff
@@ -216,10 +217,8 @@ struct ProblemDescription : ProblemDescriptionBase
 
     std::size_t GetInSize() const
     {
-        // clang-format off
-        return GetInBatchSize() * GetInChannels() * GetInDepth() * GetInHeight() *
-            GetInWidth() * GetInElementSize();
-        // clang-format on
+        return GetInBatchSize() * GetInChannels() * GetInDepth() * GetInHeight() * GetInWidth() *
+               GetInElementSize();
```
Comment (atamazov) [inline src/include/miopen/conv/problem_description.hpp:221]: [R][Quality] I guess we can use available tensor's method

File: src/include/miopen/conv/problem_description.hpp [hunk 3]
```diff
@@ -387,37 +390,41 @@ struct ProblemDescription : ProblemDescriptionBase
 #if MIOPEN_ENABLE_SQLITE
     static std::string table_name() { return "config"; }
 
-    template <class Self, class F>
-    static void Visit(Self&& self, F f)
+    template <class Self>
+    static void Visit(Self&& self, std::function<void(int, std::string)> f)
```
Comment (atamazov) [inline src/include/miopen/conv/problem_description.hpp:394]: [Q] I am afraid that changing this function breaks compatibility with existing databases.

Or was this function never used but _will_ be used after this PR instead of `ProblemDescription::Visit()`?

File: src/include/miopen/fusion/problem_description.hpp [hunk 4]
```diff
@@ -95,25 +95,26 @@ struct FusionDescription
...
-                                              conv_op.filter_desc,
-                                              out_desc,
-                                              conv_op.base_desc /* conv desc */,
-                                              dir};
+            return miopen::conv::ProblemDescription{conv_op.input_desc,
+                                                    conv_op.filter_desc,
+                                                    out_desc,
+                                                    conv_op.base_desc /* conv desc */,
+                                                    dir,
+                                                    do_bias ? 1 : 0};
```
Comment (atamazov) [inline src/include/miopen/fusion/problem_description.hpp]: [R][clarity] Can we make `conv::ProblemDescription` to accept bool here?

File: src/include/miopen/mlo_internal.hpp [hunk 5]
```diff
@@ -328,31 +290,15 @@ struct mlo_construct_base
...
-    int mloBuildConf_Key(std::string& conf_key) const
-    {
-        return _problem.mloBuildConf_Key(conf_key);
-    }
-
     std::string db_path() const { return _db_path != nullptr ? _db_path : _ctx.GetPerfDbPath(); }
 
 protected:
-    miopen::ProblemDescription _problem;
+    miopen::ProblemDescriptionCompat _problem;
```
Comment (atamazov) [inline src/include/miopen/mlo_internal.hpp]: [R][clarity]
```suggestion
    miopen::ProblemDescriptionCompatTemporary _problem;
```

File: src/solver/conv_asm_1x1u.cpp [hunk 6]
```diff
@@ -431,7 +439,16 @@ void PerformanceConfigConvAsm1x1U::RunParmeterPredictionModel(const ConvolutionC
...
 {
-    const auto elements_in_dword = 4 / GetTypeSize(problem.GetInDataType());
+    const auto in_data_type = problem.GetInDataType();
+    /// Clang-Tidy assumption:
+    /// (in_data_type == miopenDouble)  ==>
+    /// (elements_in_dword == 0) ==>
+    /// (c_mult == 0) ==>
+    /// division by zero
+    if(in_data_type == miopenDouble)
+        MIOPEN_THROW(miopenStatusInternalError);
```
Comment (atamazov) [inline src/solver/conv_asm_1x1u.cpp]: [R]
```suggestion
    if(in_data_type == miopenDouble) // For clang-tidy.
        MIOPEN_THROW(miopenStatusInternalError);
```

File: src/solver/conv_asm_1x1u.cpp [hunk 7]
```diff
@@ -431,7 +439,16 @@ void PerformanceConfigConvAsm1x1U::RunParmeterPredictionModel(const ConvolutionC
...
+
+    const auto elements_in_dword = 4 / GetTypeSize(in_data_type);
```
Comment (atamazov) [inline src/solver/conv_asm_1x1u.cpp]: [R]
```suggestion
    const auto elements_in_dword = 4 / GetTypeSize(in_data_type);
    if(elements_in_dword  == 0) // For clang-tidy.
        MIOPEN_THROW(miopenStatusInternalError);
```

File: src/solver/conv_asm_1x1u.cpp [hunk 8]
```diff
@@ -559,7 +576,7 @@ bool ConvAsm1x1U::IsApplicable(const ConvolutionContext& ctx,
         && problem.GetDilationW() == 1
         && problem.GetDilationH() == 1
         && problem.GetBias() == 0
-        && problem.GetInChannels() % elements_in_dword == 0
+        && problem.GetInChannels() % elements_in_dword == 0  // NOLINT
```
Comment (atamazov) [inline src/solver/conv_asm_1x1u.cpp]: [R] ...and revert this

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp [hunk 9]
```diff
@@ -644,6 +644,15 @@ void PerformanceConfigAsmImplicitGemmGTCWrwXdlopsNHWC::HeuristicInit(
...
     {
+        /// \todo Fix this
+        /// Clang-Tidy assumption:
+        /// (n_per_block == 0)  ==>
+        /// (n_per_block == config.gemm_n_per_block) ==>
+        /// (config.gemm_n_per_block == 0)  ==>
+        /// division by zero in GetImplicitGemmGtcDynamicWrwXdlopsNHWCKernel()
+        if(n_per_block == 0)
+            MIOPEN_THROW(miopenStatusInternalError);
+
```
Comment (atamazov) [inline src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp]: [R]
```suggestion
        // clang-tidy: DIV/0 in GetImplicitGemmGtcDynamicWrwXdlopsNHWCKernel()
        if(n_per_block == 0)
            MIOPEN_THROW(miopenStatusInternalError);

```

File: test/gtest/kernel_tuning_net.cpp [hunk 10]
```diff
@@ -46,11 +46,16 @@ struct KernelTuningNetTest : public ::testing::TestWithParam<KernelTuningNetTest
...
         tensor<G> output_tensor =
             tensor<G>(test_case.data_type, test_case.layout, test_case.output);
-        problem        = miopen::ProblemDescription(input_tensor.desc,
+
+        const bool is_forward = (test_case.direction == miopen::conv::Direction::Forward);
+        const auto conv_problem =
+            miopen::conv::ProblemDescription(is_forward ? input_tensor.desc : output_tensor.desc,
                                              weight_tensor.desc,
-                                             output_tensor.desc,
+                                             is_forward ? output_tensor.desc : input_tensor.desc,
```
Comment (atamazov) [inline test/gtest/kernel_tuning_net.cpp]: [Q] Is this change temporary?

File: src/include/miopen/conv/problem_description.hpp (hunk 3 above)
Comment (averinevg) [reply src/include/miopen/conv/problem_description.hpp:394]: This function was never used and will be used instead of `ProblemDescription::Visit()`

File: src/conv/problem_description.cpp (hunk 1 above)
Comment (averinevg) [reply src/conv/problem_description.cpp]: @atamazov I have moved this function from `src/problem_description.cpp` as is.
Don't we need to check `GetSpatialDims()` here?

File: src/include/miopen/fusion/problem_description.hpp (hunk 4 above)
Comment (averinevg) [reply src/include/miopen/fusion/problem_description.hpp]: I guess it would be even better to remove `do_bias` and use `bias` instead of it.

File: src/solver/conv_asm_1x1u.cpp (hunk 8 above)
Comment (averinevg) [reply src/solver/conv_asm_1x1u.cpp]: In this case it is obviously a false positive

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp (hunk 9 above)
Comment (averinevg) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp]: This one is useful and can be fixed, so a little explanation would be nice.

File: test/gtest/kernel_tuning_net.cpp (hunk 10 above)
Comment (averinevg) [reply test/gtest/kernel_tuning_net.cpp]: No, that's how it's meant to be. Is there something wrong?

File: src/include/miopen/conv/problem_description.hpp (hunk 2 above)
Comment (averinevg) [reply src/include/miopen/conv/problem_description.hpp:221]: Here and in several other places it would be nice to refactor, I suggest doing it in another PR

File: src/conv/problem_description.cpp (hunk 1 above)
Comment (atamazov) [reply src/conv/problem_description.cpp]: I do not know ;/

File: src/include/miopen/conv/problem_description.hpp (hunk 2 above)
Comment (atamazov) [reply src/include/miopen/conv/problem_description.hpp:221]: Resolved.

File: src/include/miopen/conv/problem_description.hpp (hunk 3 above)
Comment (atamazov) [reply src/include/miopen/conv/problem_description.hpp:394]: Resolved.

File: src/include/miopen/fusion/problem_description.hpp (hunk 4 above)
Comment (atamazov) [reply src/include/miopen/fusion/problem_description.hpp]: Up to you. Resolved.

File: src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp (hunk 9 above)
Comment (atamazov) [reply src/solver/conv_asm_implicit_gemm_gtc_wrw_nhwc.cpp]: [R] Let's try to stay brief. See updated suggestion above.

File: src/conv/problem_description.cpp (hunk 1 above)
Comment (averinevg) [reply src/conv/problem_description.cpp]: @atamazov I've changed it as you proposed and kept this check

File: test/gtest/kernel_tuning_net.cpp (hunk 10 above)
Comment (atamazov) [reply test/gtest/kernel_tuning_net.cpp]: I do not know, probably not. But if so, then why we need to use ugly code? Hmm...

**Review by atamazov — CHANGES_REQUESTED** [review]
Oh no.

File: .gitmodules [hunk 11]
```diff
@@ -1,4 +1,4 @@
 [submodule "fin"]
     path = fin
-    url = https://github.com/ROCmSoftwarePlatform/MIFin.git
+    url = https://github.com/averinevg/MIFin.git
```
Comment (atamazov) [inline .gitmodules]: This is something that needs explanation.

**Review by atamazov — CHANGES_REQUESTED** [review]
Oops

File: .gitmodules (hunk 11 above)
Comment (averinevg) [reply .gitmodules]: I'll remove this as soon as https://github.com/ROCmSoftwarePlatform/MIFin/pull/83 is merged

File: Jenkinsfile [hunk 12]
```diff
@@ -490,7 +490,7 @@ pipeline {
             description: "")
         booleanParam(
             name: "PERF_TEST",
-            defaultValue: true,
```
Comment (averinevg) [inline Jenkinsfile:493]: Performance tests have been temporary disabled

File: Jenkinsfile [hunk 13]
```diff
@@ -490,7 +490,7 @@ pipeline {
...
+            defaultValue: false,
```
Comment (averinevg) [inline Jenkinsfile]: Performance tests have been temporary disabled

**Review by DrizztDoUrden — DISMISSED** [review]
lgtm

**Review by DrizztDoUrden — DISMISSED** [review]
lgtm

**Review by DrizztDoUrden — APPROVED** [review]

File: test/sqlite_perfdb.cpp [hunk 14]
```diff
@@ -95,7 +95,7 @@ static Random& Rnd()
 
 struct ProblemData : SQLiteSerializable<ProblemData>
 {
-    ProblemDescription prob;
+    ProblemDescriptionCompatTemporary prob;
```
Comment (atamazov) [inline test/sqlite_perfdb.cpp:98]: [Future] This leftover must be removed with high urgency.

**Review by atamazov — APPROVED** [review]
LGTM!

**Comment by atamazov** [comment]
Let me run my set of performance tests before merging

**Comment by averinevg** [comment]
Today we discussed this PR with @atamazov. All validation tests passed, so we can merge the PR. Estimated time for performance test results - the day after tomorrow

**Review by junliume — APPROVED** [review]

**Comment by atamazov** [comment]
@averinevg @junliume 

## 🌀 Performance testing results

Preconditions:
- Radeon VII (gfx906/60)
- ROCm 5.4.1, Ubuntu 20.04.5
- BUILD_DEV=On, Release build
- branch 18d368c27 vs. develop ad809f9da
- 93 known FP32 convolution configs
- 93 known FP16 convolution configs

Tested modes:
- (0) Immediate mode with default (system) find-db
- (1) Find Mode (DYNAMIC_HYBRID)
- (2) Find Mode (NORMAL)
- (3) Immediate mode with user find-db filled by step (2).

Bottom line: 🟢 No performance or correctness regressions.

Logs & csv files available upon request.

File: src/problem_description.cpp [hunk 15]
```diff
@@ -8,156 +8,24 @@
...
-
-ProblemDescription::ProblemDescription(const TensorDescriptor& in,
-                                       const TensorDescriptor& weights,
-                                       const TensorDescriptor& out,
-                                       const ConvolutionDescriptor& conv,
-                                       conv::Direction dir,
-                                       int bias_)
-    : ProblemDescription(dir == conv::Direction::Forward
-                             ? conv::ProblemDescription{in, weights, out, conv, dir, bias_}
-                             : conv::ProblemDescription{out, weights, in, conv, dir, bias_})
```
Comment (atamazov) [inline src/problem_description.cpp:109]: After this PR, we do not swap xDesc and yDesc in the ProblemDescription, i.e. GetIn() always returns xDesc.

Comment (atamazov) [reply src/problem_description.cpp:109]: Oh no, we still swap descriptors.

/cc @averinevg
